feat(sdk-coin-apt): send many: addition of recipients (single recipient)

baltiyal opened the pull request:
TICKET: COIN-3327

This is part of PRs for send many support for Aptos.

- Added a new `recipients` array, currently containing a single recipient.
- Replaced internal (protected) `recipient` field with `recipients`.
- Ensured backward compatibility for existing recipient-based flows.
- Future PRs will extend `recipients` to support batch transactions with multiple recipients.

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

bhavidhingra commented on modules/sdk-coin-apt/src/lib/iface.ts:
```diff
@@ -15,7 +15,7 @@ export interface AptTransactionExplanation extends BaseTransactionExplanation {
   id: string;
   sender: string;
-  recipient: TransactionRecipient;
+  recipients: TransactionRecipient[];
```
this is a breaking change, in order to keep the code backward compatible don't modify the publicly exposed fields and methods. (Open/Closed principle i.e. open for extension, closed for modification).

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -90,12 +90,12 @@ export abstract class Transaction extends BaseTransaction {
 
-  get recipient(): TransactionRecipient {
-    return this._recipient;
+  get recipients(): TransactionRecipient[] {
```
don't modify publicly exposed methods.

bhavidhingra reviewed (changes requested):
In order to maintain backward compatible of the SDK code, pls don't modify the publicly exposed fields/methods, rather extend the code with new fields/methods. ([Open/Closed principle](https://blog.algomaster.io/i/142656403/o-openclosed-principle-ocp) i.e. open for extension, closed for modification).

at31416 commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -302,7 +306,7 @@ export abstract class Transaction extends BaseTransaction {
       id: this.id,
       sender: this.sender,
-      recipient: this.recipient,
+      recipients: this.recipients,
```
if we are modifying toJson(), we might need to modify how output of toJson is handled in WP, can you verify if this will break WP?

baltiyal replied:
`toJson()` is used fetch the `TxData` from the Transaction. As have introduced a new field `recipients`, so it should be a concern now.

baltiyal committed:
feat(sdk-coin-apt): send many: addition of recipients

TICKET: COIN-3327

- Added a new `recipients` array, currently containing a single recipient.
- Replaced internal (protected) `recipient` field with `recipients`.
- Ensured backward compatibility for existing recipient-based flows.
- Future PRs will extend `recipients` to support batch transactions with multiple recipients.

BREAKING CHANGE: Replaced protected `recipient` field with `recipients` in Aptos transactions.

at31416 commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -98,6 +99,14 @@ export abstract class Transaction extends BaseTransaction {
+    return this._recipients;
+  }
+
+  set recipients(value: TransactionRecipient[]) {
```
should we set `recipient` field as well with the value of first item of the array? I'm asking because if we make changes in WP to set `recipients` instead of `recipient` will other methods which return `recipient` value return `null` for `recipient`?

baltiyal replied:
I don't think we would need to do that. As when migrating from WP `recipient` to `recipients` these changes are backward compatible so won't break anything. Still, lets discuss once on this.

at31416 commented on modules/sdk-coin-apt/test/unit/transactionBuilder/fungibleAssetTransferBuilder.ts:
```diff
@@ -8,169 +8,341 @@ import { FungibleAssetTransfer } from '../../../src/lib/transaction/fungibleAsse
+        const tx = (await txBuilder.build()) as FungibleAssetTransfer;
+        const toJson = tx.toJson();
+        should.equal(toJson.sender, testData.sender2.address);
+        should.deepEqual(toJson.recipients, [
```
check for `recipient` as well here and in other similar places

baltiyal replied:
Same as above.

baltiyal commented on modules/sdk-coin-apt/test/unit/transactionBuilder/digitalAssetTransferBuilder.ts:
```diff
@@ -7,138 +7,280 @@ import { DigitalAssetTransfer } from '../../../src/lib/transaction/digitalAssetT
+      const tx = (await txBuilder.build()) as DigitalAssetTransfer;
+      const toJson = tx.toJson();
+      should.equal(toJson.id, '0x3a97bbf538a73f98625b65c770bbf69d032ac18cd028b871182a127f10ab5666');
+      should.equal(toJson.sender, testData.sender2.address);
```
`Recipient` and `Recipients` are two separate flows, we have recipient test case in backward compatibility describe block. These flows are mutually exclusive to each other. Still lets discuss once, to be exhaustive.

baltiyal commented on modules/sdk-coin-apt/test/unit/transactionBuilder/digitalAssetTransferBuilder.ts:
```diff
@@ -7,138 +7,280 @@ import { DigitalAssetTransfer } from '../../../src/lib/transaction/digitalAssetT
+      const tx = (await txBuilder.build()) as DigitalAssetTransfer;
+      const toJson = tx.toJson();
+      should.equal(toJson.sender, testData.sender2.address);
+      should.deepEqual(toJson.recipients, [
```
Same as above.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -39,6 +39,7 @@ export abstract class Transaction extends BaseTransaction {
   protected _feePayerSignature: Signature;
   protected _sender: string;
   protected _recipient: TransactionRecipient;
```
please remove this field

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/digitalAssetTransfer.ts:
```diff
@@ -43,13 +43,21 @@ export class DigitalAssetTransfer extends Transaction {
     const senderAddress = AccountAddress.fromString(this._sender);
-    const recipientAddress = AccountAddress.fromString(this._recipient.address);
+    const recipientAddress = AccountAddress.fromString(
+      this.recipients ? this._recipients[0].address : this._recipient.address
```
all private and protected should use `recipients` because `_recipient` will be removed in this PR.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -39,6 +39,7 @@ export abstract class Transaction extends BaseTransaction {
   protected _feePayerSignature: Signature;
   protected _sender: string;
   protected _recipient: TransactionRecipient;
+  protected _recipients: TransactionRecipient[];
```
initialized `_recipients` with empty list so that we don't have to add null check anywhere.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -258,14 +267,24 @@ export abstract class Transaction extends BaseTransaction {
   }
 
   loadInputsAndOutputs(): void {
+    const totalAmount = this._recipients?.reduce(
```
`_recipients` will be initialized with empty list in the constructor.
```suggestion
    const totalAmount = this._recipients.reduce(
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -258,14 +267,24 @@ export abstract class Transaction extends BaseTransaction {
       {
         address: this.sender,
-        value: this.recipient.amount as string,
+        value: totalAmount?.toString() || (this.recipient.amount as string),
```
```suggestion
        value: totalAmount.toString(),
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -258,14 +267,24 @@ export abstract class Transaction extends BaseTransaction {
       },
     ];
-    this._outputs = [
+    this._outputs = this._recipients?.map((recipient) => {
```
remove null checks from all places because `_recipients` will be initialized with empty list in the constructor.
```suggestion
    this._outputs = this._recipients.map((recipient) => {
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transactionBuilder/transactionBuilder.ts:
```diff
@@ -70,6 +70,13 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
   }
 
+  recipients(recipients: Recipient[]): this {
+    this.validateAddress({ address: recipients[0].address });
```
do validations in a for loop for all recipients.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transactionBuilder/transactionBuilder.ts:
```diff
@@ -143,8 +150,13 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
     this.validateAddress({ address: transaction.sender });
-    this.validateAddress({ address: transaction.recipient.address });
-    this.validateValue(new BigNumber(transaction.recipient.amount));
+    if (transaction.recipients) {
```
please do validations in a for loop for all recipients.

bhavidhingra commented on modules/sdk-coin-apt/test/unit/transactionBuilder/digitalAssetTransferBuilder.ts:
```diff
@@ -7,138 +7,280 @@ import { DigitalAssetTransfer } from '../../../src/lib/transaction/digitalAssetT
-    txBuilder.gasData({
-      maxGasAmount: 200000,
-      gasUnitPrice: 100,
+  describe('Backward Compatibility with Recipient', () => {
```
please prevent code duplicity.
remove the two different set of tests i.e. `Backward Compatibility with Recipient` and `Forward Compatibility with Recipients`.
and extend the existing tests only.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -335,8 +355,10 @@ export abstract class Transaction extends BaseTransaction {
 
-    const outputs: TransactionRecipient[] = [this.recipient];
-    const outputAmount = outputs[0].amount;
+    const outputs: TransactionRecipient[] = this._recipients || [this._recipient];
```
```suggestion
    const outputs: TransactionRecipient[] = [this._recipient];
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -335,8 +355,10 @@ export abstract class Transaction extends BaseTransaction {
 
-    const outputs: TransactionRecipient[] = [this.recipient];
-    const outputAmount = outputs[0].amount;
+    const outputs: TransactionRecipient[] = this._recipients || [this._recipient];
+    const outputAmount = outputs
+      .reduce((accumulator, current) => accumulator.plus(current.amount), new BigNumber('0'))
+      .toString();
```
```suggestion
    const outputs: TransactionRecipient[] = [this.recipient];
    const outputAmount = outputs[0].amount;
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/digitalAssetTransfer.ts:
```diff
@@ -43,13 +43,21 @@ export class DigitalAssetTransfer extends Transaction {
     const aptos = new Aptos(new AptosConfig({ network }));
     const senderAddress = AccountAddress.fromString(this._sender);
-    const recipientAddress = AccountAddress.fromString(this._recipient.address);
+    const recipientAddress = AccountAddress.fromString(
+      this.recipients ? this._recipients[0].address : this._recipient.address
+    );
```
```suggestion
    const recipientAddress = AccountAddress.fromString(this.recipient.address);
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/fungibleAssetTransfer.ts:
```diff
@@ -38,15 +38,23 @@ export class FungibleAssetTransfer extends Transaction {
     const aptos = new Aptos(new AptosConfig({ network }));
     const senderAddress = AccountAddress.fromString(this._sender);
-    const recipientAddress = AccountAddress.fromString(this._recipient.address);
+    const recipientAddress = AccountAddress.fromString(
+      this._recipients ? this._recipients[0].address : this._recipient.address
+    );
```
```suggestion
    const recipientAddress = AccountAddress.fromString(this.recipient.address);
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/fungibleAssetTransfer.ts:
```diff
@@ -57,7 +65,7 @@ export class FungibleAssetTransfer extends Transaction {
         function: FUNGIBLE_ASSET_TRANSFER_FUNCTION,
         typeArguments: [FUNGIBLE_ASSET_TYPE_ARGUMENT],
-        functionArguments: [fungibleTokenAddress, recipientAddress, this.recipient.amount],
+        functionArguments: [fungibleTokenAddress, recipientAddress, recipientAmount],
```
```suggestion
        functionArguments: [fungibleTokenAddress, recipientAddress, this.recipient.amount],
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/fungibleAssetTransfer.ts:
```diff
@@ -57,7 +57,7 @@ export class FungibleAssetTransfer extends Transaction {
         function: FUNGIBLE_ASSET_TRANSFER_FUNCTION,
         typeArguments: [FUNGIBLE_ASSET_TYPE_ARGUMENT],
-        functionArguments: [fungibleTokenAddress, recipientAddress, this.recipient.amount],
+        functionArguments: [fungibleTokenAddress, recipientAddress, this.recipients[0].amount],
```
```suggestion
        functionArguments: [fungibleTokenAddress, recipientAddress, this.recipient.amount],
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/digitalAssetTransfer.ts:
```diff
@@ -37,19 +37,20 @@ export class DigitalAssetTransfer extends Transaction {
     const aptos = new Aptos(new AptosConfig({ network }));
     const senderAddress = AccountAddress.fromString(this._sender);
-    const recipientAddress = AccountAddress.fromString(this._recipient.address);
+    const recipientAddress = AccountAddress.fromString(this._recipients[0].address);
```
nit:
```suggestion
    const recipientAddress = AccountAddress.fromString(this.recipients[0].address);
```

baltiyal replied:
Aren't both same.
As `_recipients` is a protected field, and this is a child class.

bhavidhingra replied:
yes, it is technically the same. You can a more elaborate comment about this suggestion [here].(https://github.com/BitGo/BitGoJS/pull/5809/files#r2009731697)

baltiyal replied:
Done for child classes.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/fungibleAssetTransfer.ts:
```diff
@@ -32,21 +32,21 @@ export class FungibleAssetTransfer extends Transaction {
     const aptos = new Aptos(new AptosConfig({ network }));
     const senderAddress = AccountAddress.fromString(this._sender);
-    const recipientAddress = AccountAddress.fromString(this._recipient.address);
+    const recipientAddress = AccountAddress.fromString(this._recipients[0].address);
```
nit:
```suggestion
    const recipientAddress = AccountAddress.fromString(this.recipients[0].address);
```

baltiyal replied:
Aren't both same.
As `_recipients` is a protected field, and this is a child class.

baltiyal replied:
Done.

baltiyal replied:
Done for child clases.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -91,11 +92,19 @@ export abstract class Transaction extends BaseTransaction {
 
   set recipient(value: TransactionRecipient) {
-    this._recipient = value;
+    this._recipients[0] = value;
```
```suggestion
    this.recipients = [value];
```

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -258,20 +267,24 @@ export abstract class Transaction extends BaseTransaction {
   }
 
   loadInputsAndOutputs(): void {
+    const totalAmount = this._recipients.reduce(
```
```suggestion
    const totalAmount = this.recipients.reduce(
```

baltiyal replied:
Aren't both same.
As `_recipients` is a protected field, and this is a child class.

bhavidhingra replied:
we discussed in the meet that it is a better practice to have private fields and access them using setters and getters in the child classes. We can't modify them to private fields, but using setters and getters is still possible.
Let's make it a practice of using setters and getters in child classes till we find an issue with that. No need to modify other fields in this PR, only this one.

bhavidhingra replied:
yes, correct

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -258,20 +267,24 @@ export abstract class Transaction extends BaseTransaction {
-      {
-        address: this.recipient.address,
-        value: this.recipient.amount as string,
+    this._outputs = this._recipients.map((recipient) => {
```
```suggestion
    this._outputs = this.recipients.map((recipient) => {
```

baltiyal replied:
Aren't both same.
As `_recipients` is a protected field, and this is a child class.

bhavidhingra replied:
same as above

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -91,11 +92,19 @@ export abstract class Transaction extends BaseTransaction {
 
   get recipient(): TransactionRecipient {
-    return this._recipient;
+    return this._recipients[0];
```
```suggestion
    return this._recipients.length > 0 ? this._recipients[0] : undefined;
```

baltiyal replied:
`TS2322: Type TransactionRecipient | undefined is not assignable to type TransactionRecipient
Type undefined is not assignable to type TransactionRecipient`

at31416 replied:
first we have to verify whether `_recipients` filed is not undefined/null then this error will go away i think.

baltiyal replied:
I have put an assertion for it.

bhavidhingra replied:
@at31416 `this._recipients` field is always non-null because it is initialized with empty array in the constructor.

bhavidhingra replied:
@baltiyal thanks, assertion will work.

at31416 commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -335,8 +349,10 @@ export abstract class Transaction extends BaseTransaction {
 
-    const outputs: TransactionRecipient[] = [this.recipient];
-    const outputAmount = outputs[0].amount;
+    const outputs: TransactionRecipient[] = this._recipients;
```
We can change `_recipients` to `recipients` in every place which will use getter instead of accessing the field directly

```suggestion
    const outputs: TransactionRecipient[] = this.recipients;
```

baltiyal replied:
Accessing a protected field `_recipients` in the same class. I don't think its a issue.
All the fields like `_sender`,` _maxGasAmount`, `_expirationTime` , etc. all are directly accessed, which is never a issue in the same class.

bhavidhingra replied:
I agree with @at31416

at31416 commented on modules/sdk-coin-apt/src/lib/transaction/transferTransaction.ts:
```diff
@@ -30,25 +30,26 @@ export class TransferTransaction extends Transaction {
     this._assetId = entryFunction.type_args[0].toString();
-    this._recipient.address = entryFunction.args[0].toString();
-    this._recipient.amount = utils.getAmountFromPayloadArgs(entryFunction.args[1].bcsToBytes());
+    this._recipients = [
```
This sets a single recipient always when we call `parseTransactionPayload`? even if we had multiple recipients?

baltiyal replied:
This PR is meant for single recipient only. Logic for parsing and building for multiple recipients would be added in next PR.

bhavidhingra replied:
we might have to modify this function for multiple recipients, but multi-recipient handling will be picked up in a separate PR.

bhavidhingra commented on modules/sdk-coin-apt/test/unit/transactionBuilder/digitalAssetTransferBuilder.ts:
```diff
@@ -12,7 +12,7 @@ describe('Apt Digital Asset Transfer Builder', () => {
     const txBuilder = factory.getDigitalAssetTransactionBuilder(digitalAssetTransfer);
     txBuilder.sender(testData.sender2.address);
-    txBuilder.recipient(testData.digitalTokenRecipients[0]);
+    txBuilder.recipients(testData.digitalTokenRecipients);
```
```suggestion
    txBuilder.recipient(testData.digitalTokenRecipients[0]);
    txBuilder.recipients(testData.digitalTokenRecipients);
```

bhavidhingra commented on modules/sdk-coin-apt/test/unit/transactionBuilder/digitalAssetTransferBuilder.ts:
```diff
@@ -141,4 +154,98 @@ describe('Apt Digital Asset Transfer Builder', () => {
     should.equal(toJson.expirationTime, 1738041170);
   });
+
+  describe('should test for deprecated field recipient in building transaction, without usage of recipients', () => {
```
I think we can mark those fields as deprecated also in this PR.

baltiyal replied:
Marked. Please check once is it done correctly?

bhavidhingra replied:
yes correct, just suggested a slight change taking reference from [here](https://github.com/BitGo/BitGoJS/blob/master/modules/sdk-api/src/bitgoAPI.ts#L1624).

baltiyal commented:
When adding commit message feat(sdk-coin-apt)`!` according to this : https://www.conventionalcommits.org/en/v1.0.0/ , the commit lint CI is breaking. So, have added in BREAKING CHANGE in the footer instead.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -90,12 +92,27 @@ export abstract class Transaction extends BaseTransaction {
   }
 
+  /**
+   * @deprecated Use `recipients` instead.
```
```suggestion
   * * @deprecated - use `recipients()`.
```

baltiyal replied:
Nit comment, but done.

bhavidhingra commented on modules/sdk-coin-apt/src/lib/transaction/transaction.ts:
```diff
@@ -90,12 +92,27 @@ export abstract class Transaction extends BaseTransaction {
   }
 
+  /**
+   * @deprecated Use `recipients` instead.
```
```suggestion
   * @deprecated - use `recipients()`.
```

baltiyal replied:
Nit comment, but done.

bhavidhingra reviewed (approved):
LGTM, please test it with express local setup

bhavidhingra reviewed (changes requested):
@baltiyal please modify the commit message, there is some stale information in it.

> This is part of PRs for send many support for Aptos.
> In this PR, adding new array of recipients but currently its having a single recipient in it.
> Along with the maintaining backward compatibility for recipient flow.
> In upcoming PRs, would be incrementally changing recipients to have batch support and multiple recipients.
> And after successful migration, removing/deprecating the recipient field.

at31416 commented on modules/sdk-coin-apt/src/lib/transactionBuilder/transactionBuilder.ts:
```diff
@@ -63,10 +63,22 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
     this.validateAddress({ address: recipient.address });
     this.validateValue(new BigNumber(recipient.amount));
-    this.transaction.recipient = recipient;
+    this.transaction.recipients[0] = recipient;
```
This might end up in runtime error, what do you think of this suggestion.

```suggestion
    this.transaction.recipients = [recipient];
```

bhavidhingra replied:
@at31416 I was about to give this error, but then had a thought about how the test cases is not getting runtime exception. Then I asked chatgpt and turns out that it is allowed in typescript.
